psi4/src/psi4/libfock/DFJCOSK.cc

```diff
@@ -0,0 +1,1014 @@
+
+        int rank = 0;
+#ifdef _OPENMP
+        rank = omp_get_thread_num();
+#endif
+
+        // grid points in this block
+        auto block = grid->blocks()[bi];
+        int npoints_block = block->npoints();
+        auto x = block->x();
+        auto y = block->y();
+        auto z = block->z();
+        auto w = block->w();
+
+        // significant basis functions and shells at these grid points
+        // significance determined via basis extent
+        std::vector<int> bf_map = block->functions_local_to_global();
+        std::vector<int> shell_map = block->shells_local_to_global();
+        int nbf_block = bf_map.size();
+        int ns_block = shell_map.size();
+
+        // lists of all basis functions and shells
+        //
+        // The use of these "all" lists adds O(N^2) cost to the COSK grid loop (w/ small prefactor)
+        // This cost is negligible relative to the esp integral computation, which is O(N) (w/ a much larger prefactor),
+        // but future COSK work could remove this potential bottleneck
+        std::vector<int> bf_map_all;
+        std::vector<int> shell_map_all;
+        for (size_t bf = 0; bf < nbf; bf++) bf_map_all.push_back(bf);
+        for (size_t s = 0; s < nshell; s++) shell_map_all.push_back(s);
+        int nbf_block_all = bf_map_all.size();
+        int ns_block_all = shell_map_all.size();
+
+        // => Bookkeeping <= //
+
+        // map index in shell_map_all to first index in bf_map_all
+        std::vector<int> shell_map_all_to_bf_map_all;
+        if (shell_map_all.size() > 0) {
+            shell_map_all_to_bf_map_all.push_back(0);
+            for(size_t shell_map_ind = 0; (shell_map_ind + 1) < shell_map_all.size(); shell_map_ind++) {
+                size_t MU = shell_map_all[shell_map_ind];
+                shell_map_all_to_bf_map_all.push_back(primary_->shell(MU).nfunction() + shell_map_all_to_bf_map_all.back());
+            }
+        }
+        
+        // map index in shell_map to first index in bf_map
+        std::vector<int> shell_map_to_bf_map;
+        if (shell_map.size() > 0) {
+            shell_map_to_bf_map.push_back(0);
+            for(size_t shell_map_ind = 0; (shell_map_ind + 1) < shell_map.size(); shell_map_ind++) {
+                size_t MU = shell_map[shell_map_ind];
+                shell_map_to_bf_map.push_back(primary_->shell(MU).nfunction() + shell_map_to_bf_map.back());
+            }
+        }
+
+        // map back from global shell index to local shell index
+        std::map<size_t, size_t> shell_map_inv;
+        for (size_t shell_ind = 0; shell_ind < shell_map.size(); shell_ind++) {
+            shell_map_inv[shell_map[shell_ind]] = shell_ind;
+        }
+
+        // => Process Density Matrix <= //
+
+        // significant cols of D for this grid block
+        std::vector<SharedMatrix> D_block(njk);
+        for(size_t jki = 0; jki < njk; jki++) {
+            D_block[jki] = std::make_shared<Matrix>(nbf_block_all, nbf_block);
```

Comment:
Correct me if I am wrong, but it seems that this formulation of D_block results in the COSX implementation using O(N^3) memory - D_block is a vector of size njk=D.size()=nbf*nbf containing pointers to shared matrices each of size (nbf_block_all=nbf)*nbf_block. Possibly something to defer to a future PR, but would it be feasible to change the algo so it doesn't use O(N^3) memory?

Reply:
`njk` is not `nbf ^ 2`, it's the number of unique `J` and `K` objects to be formed (1 in the RHF case, 2 in the UHF case, etc.)

Also `nbf_block` is constant in size. Thus, the total memory usage is only O(N), not O(N^3)

Reply:
Ah! You're completely right. D isn't just a single density matrix - it's the vector of individual density matrices. It would be D[i] that is nbf*nbf. And yeah, nbf_block is constant, and nbf_block_all is O(N), so you're right. Once again, my mistake!

Maybe I'm not in the right state of mind to review this code right now. I do apologize.

Reply:
All good! You're asking the right types of questions